lib/networkaccessmanager.cpp

```diff
@@ -47,16 +53,70 @@ static NetworkAccessManager* createNam()
 
 NetworkAccessManager* NetworkAccessManager::instance()
 {
-    static auto* nam = createNam();
-    return nam;
+    static QThreadStorage<NetworkAccessManager *> storage;
+    if(!storage.hasLocalData()) {
+        storage.setLocalData(createNam());
+    }
+    return storage.localData();
 }
 
 NetworkAccessManager::~NetworkAccessManager() = default;
 
 QNetworkReply* NetworkAccessManager::createRequest(
     Operation op, const QNetworkRequest& request, QIODevice* outgoingData)
 {
+    if(request.url().scheme() == QStringLiteral("mxc")) {
+        const auto fragment = request.url().fragment();
+        const auto fragmentParts = fragment.split(QLatin1Char('/'));
+        const auto mediaId = request.url().toString(QUrl::RemoveScheme | QUrl::RemoveFragment);
+        if(fragmentParts.size() == 3) {
+            auto connection = AccountRegistry::instance().get(fragmentParts[0]);
+            if(!connection) {
+                qWarning() << "Connection not found";
+                return nullptr;
+            }
+            auto room = connection->room(fragmentParts[1]);
+            if(!room) {
+                qWarning() << "Room not found";
+                return nullptr;
+            }
+            QNetworkRequest r(request);
+            r.setUrl(QUrl(QStringLiteral("%1/_matrix/media/r0/download/%2").arg(connection->homeserver().toString(), mediaId)));
+            auto reply = createRequest(QNetworkAccessManager::GetOperation, r);
+            return new MxcReply(reply, room, fragmentParts[2]);
+        } else if(fragmentParts.size() == 1) {
+            auto connection = AccountRegistry::instance().get(fragment);
+            if(!connection) {
+                qWarning() << "Connection not found";
+                return nullptr;
+            }
+            QNetworkRequest r(request);
+            r.setUrl(QUrl(QStringLiteral("%1/_matrix/media/r0/download/%2").arg(connection->homeserver().toString(), mediaId)));
+            auto reply = createRequest(QNetworkAccessManager::GetOperation, r);
+            return new MxcReply(reply);
+        } else {
+            qWarning() << "Invalid request";
+            return nullptr;
+        }
+    }
     auto reply = QNetworkAccessManager::createRequest(op, request, outgoingData);
     reply->ignoreSslErrors(d->ignoredSslErrors);
     return reply;
 }
+
+QStringList NetworkAccessManager::supportedSchemesImplementation() const
+{
+    auto schemes = QNetworkAccessManager::supportedSchemesImplementation();
+    schemes += QStringLiteral("mxc");
+    return schemes;
+}
+
+QUrl NetworkAccessManager::urlForRoomEvent(Room *room, const QString &eventId, const QString &mediaId)
+{
+    return QUrl(QStringLiteral("mxc:%1#%2/%3/%4").arg(mediaId, room->connection()->userId(), room->id(), eventId));
```

Comment:
This layout is not prepared to receive event ids (as well as historical user ids) with slashes. Consider using `toPercentEncoding()` (and AFAIR `QUrl::fragment()` will do the right thing and decode slashes back automatically).